### DISCUSSION TITLE
Include "experimental" `DecodeHints::tryDenoise` feature in next release?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#577] [Polls]
Thanks to @Sergio- 's donation I recently [added](https://github.com/zxing-cpp/zxing-cpp/commit/9bcfdb3637714af9419bae99dd4c48a12625ab1e) a new `tryDenoise` flag to `DecodeHints` to improve the detection of matrix symbologies in the presence of noise. The use case was noise introduced by unreliable print quality of retail point of sale thermal printers. Here is a sample to get an idea what I'm talking about:
![noisy-qr](https://github.com/zxing-cpp/zxing-cpp/assets/1038346/e9a67f5e-cf1f-490c-9754-c782c4b0aac3)

Now I'm wondering whether or not to disable that new flag prior to the next release due to lack of real-world feedback so far. That is because I don't want to add half-baked stuff that I then can't easily change/remove again (see the whole #333 discussion).

**Comment by axxel** [comment]
The lack of feedback convinced me to disable this feature before the next release. See the new `BUILD_EXPERIMENTAL_API` cmake option [here](https://github.com/zxing-cpp/zxing-cpp/commit/c604fb6c91a8947ac45676b3cc478f4bc4e21c04).

**Comment by bansal21ankit** [comment]
@axxel can you help me understand how to enable this `BUILD_EXPERIMENTAL_API` flag to check this `tryDenoise` hint ? I am trying to validate it using the WASM wrapper.

**Reply by axxel** [reply]
3 things need to be taken care of in this case:
 * use the current HEAD because the 2.1 release has an issue regarding this "feature"
 * configure your project with the `cmake` command line option  `-DBUILD_EXPERIMENTAL_API=ON`
 * modify the [BarcodeReader.cpp](https://github.com/zxing-cpp/zxing-cpp/blob/63dda1c2aa43923825f337b47fe98c84b9c39a6a/wrappers/wasm/BarcodeReader.cpp#L37) by adding the line `hints.setTryDenoise(true);`

**Reply by bansal21ankit** [reply]
Thanks, I tried removing `BUILD_EXPERIMENTAL_API` check from the code and also turned `ON` the flag in `CMakeList` files. The `tryDenoise` hint **does** make a difference in detection speed and quality.

But even this is not able to detect codes with some impurities.
Case : Sometimes QR codes need to be printed on invoices, tickets, etc. having a pre-printed template like lines or blocks. These codes can get overlapped with these template items (like one of the sample here)
![qr_code_big_impure](https://github.com/zxing-cpp/zxing-cpp/assets/5683899/6da6823c-9654-425d-ab3f-175b40f31064)

**Reply by axxel** [reply]
This red line has nothing to do with 'noise'. It simply destroys 2 of 3 finder patterns and they are the features that make QRCodes recognizable as such. This image is not decodeable with zxing-cpp and will unlikely be anytime soon. My advice: don't break your symbol by printing random lines over them.
